darwin/dataset/upload_manager.py

```diff
@@ -43,72 +31,194 @@ def add_files_to_dataset(
+            payload = {"image_filenames": [image.name for image in images],
+                       "videos": [{"fps": fps, "original_filename": video.name}
+                                  for video in videos]}
         )
+        if 'errors' in data:
+            raise ValueError(f"There are errors in the put request: {data['errors']['detail']}")
+
+        if images:
+            g = (lambda images: (
+                functools.partial(
+                    _delayed_upload_function,
+                    client=client,
+                    file=image_file,
+                    files_path=images,
+                    endpoint_prefix="dataset_images"
+                )
+                for image_file in data["image_data"]
+            ))
+            generators.append(g(images))
+
+        if videos:
+            g = (lambda videos: (
+                functools.partial(
+                    _delayed_upload_function,
+                    client=client,
+                    file=video_file,
+                    files_path=videos,
+                    endpoint_prefix="dataset_videos"
+                )
+                for video_file in data["video_data"]
+            ))
+            generators.append(g(videos))
+
+    assert generators
+    return itertools.chain(*generators), len(filenames)
+
+
+def _split_on_file_type(files: List[Path]):
+    """Splits a single list of files into images and videos based on their extension
 
-        generator = itertools.chain(image_generator(), video_generator())
+    Parameters
+    ----------
+    files : list[Path]
+        List of files to split according to their type
 
-        return generator, len(filenames)
+    Returns
+    -------
+    images, videos : list[Path]
+        List of image and videos, respectively
+    """
+    images = []
+    videos = []
+    for file_path in files:
+        suffix = file_path.suffix
+        if suffix in SUPPORTED_IMAGE_EXTENSIONS:
+            images.append(file_path)
+        elif suffix in SUPPORTED_VIDEO_EXTENSIONS:
+            videos.append(file_path)
+        else:
+            raise UnsupportedFileType(file_path)
+    return images, videos
 
 
-def chunk(items, size):
+def _chunk_filenames(files: List[Path], size: int):
     """ Chunks paths in batches of size.
     No batch has any duplicates with regards to file name.
     This is needed due to a limitation in the upload api.
+
+    Parameters
+    ----------
+    files : list[Path]
+        List of files to chunk
+    size : int
+        Chunk size
+
+    Returns
+    -------
+        Chunk of the list with the next `size` elements from `files`
     """
     current_list = []
     current_names = set()
     left_over = []
-    for item in items:
-        name = Path(item).name
-        if name in current_names:
-            left_over.append(item)
+    for file in files:
+        if file.name in current_names:
+            #TODO I don't fully understand this logic. We enter here IFF there is the same
```

Comment:
I think it's worth removing the `TODO` from the code, and ask it to the rest of the team, so we make sure everybody is on the same page

Reply:
This code is a bit weird but this is the general gist of it:

We send the filenames to the backend before we start uploading the data, if we sent the full file path then it's easy to know which dataset_image_id belongs to which file on disk (we get a list of {dataset_image_id and full_file_path back). 
But we don't want the full path to be sent to the server since it will be displayed in the UI and users won't like that .

So we need to just send the filename, but then if we have more than one file with that filename  (in the same upload: e.g. /cat/1.png and /dot/1.png (since both are 1.png)) we don't know which dataset_image_id goes to which image. 

We solved that by making sure they they aren't chunked together, e.g. /cat/1.png and /dog/1.png will go in two different chunks. 

There is probably a better solution to this :)